### PR TITLE
Implement RuntimeError::QueryForUndefinedRule

### PR DESCRIPTION
**Pull request by patrickod**
Cut a RuntimeError::UndefinedRuleError and throw it in query_for_predicate when
there are no rule entries for the given name. Previously this backtracked
silently.

PR checklist:

- [ ] Added changelog entry.

### CODE REVIEW

**Review by gj — CHANGES_REQUESTED** [review]

File: languages/ruby/spec/oso/polar/polar_spec.rb [hunk 1]
```diff
@@ -157,7 +157,7 @@ def initialize
       subject.load_file(test_file)
       expect(qvar(subject, 'f(x)', 'x')).to eq([1, 2, 3])
       subject.clear_rules
-      expect(query(subject, 'f(x)')).to eq([])
+      expect { query(subject, 'f(x)') }.to raise_error Oso::Polar::PolarRuntimeError
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/ruby/spec/oso/polar/polar_spec.rb]: b5653187

File: languages/python/django-oso/tests/test_django_oso.py [hunk 2]
```diff
@@ -234,6 +234,9 @@ def test_partial_errors(rf):
     request = rf.get("/")
     request.user = "test_user"
 
+    # unrelated allow rule to prevent runtime error for query to undefined #
+    # rule.
+    load_additional_str('allow("foo", "bar", "baz");')
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/python/django-oso/tests/test_django_oso.py]: Updated to mirror (what I think is) the corresponding test in `sqlalchemy-oso`: 157aa67c

File: polar-core/src/error.rs [hunk 3]
```diff
@@ -433,6 +438,7 @@ The expression is: {expr}
             // Refactor.
             Self::InvalidState { msg } => write!(f, "Invalid state: {}", msg),
             Self::MultipleLoadError => write!(f, "Cannot load additional Polar code -- all Polar code must be loaded at the same time."),
+            Self::UndefinedRuleError { name } => write!(f, "Cannot evaluate query for undefined rule `{}`", name),
```
[the remark this answers is not in the document]
Comment (gj) [reply polar-core/src/error.rs]: 9ab5b666

File: polar-core/src/error.rs [hunk 4]
```diff
@@ -299,6 +300,9 @@ pub enum RuntimeError {
         msg: String,
     },
     MultipleLoadError,
+    UndefinedRuleError {
```
[the remark this answers is not in the document]
Comment (gj) [reply polar-core/src/error.rs]: a7745795

File: languages/go/tests/unit_test.go [hunk 5]
```diff
@@ -100,16 +100,12 @@ func TestClearRules(t *testing.T) {
...
-		var got []map[string]interface{}
-		for elem := range results {
-			got = append(got, elem)
-		}
-		if len(got) > 0 {
-			t.Errorf("Received too many results: %v", got)
-		}
+	if err = <-errors; err == nil {
+		t.Error("Expected query for undefined rule to throw error")
+	}
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/go/tests/unit_test.go:105]: 93617476

File: languages/go/tests/unit_test.go [hunk 6]
```diff
@@ -194,12 +190,12 @@ func TestQueryRule(t *testing.T) {
 	}
 
 	results, errors = o.QueryRule("v", 1)
+	if e := <-errors; e == nil {
+		t.Error("Expected query for undefined rule to throw error")
+	}
```
[the remark this answers is not in the document]
Comment (gj) [reply languages/go/tests/unit_test.go]: 93617476

**Review by samscott89 — APPROVED** [review]

**Review by gj — APPROVED** [review]

**Pull request merged**